src/NLog.Extensions.Logging/Logging/NLogLogger.cs

```diff
@@ -55,33 +65,126 @@ public void Log<TState>(Microsoft.Extensions.Logging.LogLevel logLevel, EventId
             {
-                SetLogEventMessageFormatter(logEvent, messageTemplateParameters, message);
+                for (int i = 0; i < messagetTemplateParameters.Count; ++i)
+                {
+                    if (messagetTemplateParameters[i].Name != messageParameters[i].Name)
+                    {
+                        return CreateLogEventInfoParametersSlow(messageParameters, messagetTemplateParameters);
+                    }
+                }
+
+                // Everything is mapped correctly, inject messageParameters directly as params-array
+                var paramsArray = new object[messagetTemplateParameters.Count];
+                for (int i = 0; i < paramsArray.Length; ++i)
+                    paramsArray[i] = messageParameters[i].Value;
+                return paramsArray;
+            }
+
+            return CreateLogEventInfoParametersSlow(messageParameters, messagetTemplateParameters);
+        }
+
+        /// <summary>
+        /// Resolves mismatch between the input from Microsoft Extension Logging TState and NLog Message Template Parser
+        /// </summary>
+        private object[] CreateLogEventInfoParametersSlow(NLogMessageParameterList messageParameters, NLog.MessageTemplates.MessageTemplateParameters messagetTemplateParameters)
+        {
+            if (messagetTemplateParameters.IsPositional)
+            {
+                // Find max number
+                int maxIndex = 0;
+                for (int i = 0; i < messagetTemplateParameters.Count; ++i)
+                {
+                    if (messagetTemplateParameters[i].Name.Length == 1)
+                        maxIndex = Math.Max(maxIndex, messagetTemplateParameters[i].Name[0] - '0');
+                    else
+                        maxIndex = Math.Max(maxIndex, int.Parse(messagetTemplateParameters[i].Name));
+                }
+                var paramsArray = new object[maxIndex + 1];
+                for (int i = 0; i < messageParameters.Count; ++i)
+                {
+                    // First positional name is the startPos
+                    if (char.IsDigit(messagetTemplateParameters[i].Name[0]))
+                    {
+                        for (int j = 0; j <= maxIndex; ++j)
+                        {
+                            if (i + j < messageParameters.Count)
+                                paramsArray[j] = messageParameters[i + j].Value;
+                        }
+                        break;
+                    }
+                }
+                return paramsArray;
+            }
+            else
+            {
+                var paramsArray = new object[messagetTemplateParameters.Count];
+                int startPos = 0;
+                for (int i = 0; i < messagetTemplateParameters.Count; ++i)
+                {
+                    for (int j = startPos; j < messageParameters.Count; ++i)
```

Comment:
it would be nice if we could refactor a bit, so we haven't nested `for` loops in one method

Reply:
Well it merges the two lists, so it would be awkward not to have both lists, but one could split positional and non-positional.